File: Problem3/Solution.java
import java.util.Map;
import java.util.HashMap;


class Solution {

    /**
     * Time complexity: O(n).
     * Runtime: 20 ms, faster than 88.74% of Java online submissions for Longest Substring Without Repeating Characters.
     * Memory Usage: 39.1 MB, less than 23.93% of Java online submissions for Longest Substring Without Repeating Characters.
     *
     * @param s String to examine
     * @return length of longest non-recurring substring
     * @see Integer
     */
    public Integer lengthOfLongestSubstring(String s) {

        Map<Character, Integer> hashTable = new HashMap<>();
        int i = 0, j = 0, len = 0;

        while (j < s.length()) {
            if (hashTable.containsKey(s.charAt(j))) {
                // slide i
                i = Math.max(i, hashTable.get(s.charAt(j)) + 1);
            }

            hashTable.put(s.charAt(j), j++);

            // sliding window indexes, [i, j)
            if (len < j - i) len = j - i;
        }

        return len;
    }
}

File: Problem3/Solution.py
class Solution:

    def lengthOfLongestSubstring(self, s):

        """
        Time complexity: O(n).

        args:
            s: str

        returns: int, length of longest non-recurring substring.
        """

        hashtable = dict()
        l = len(s)

        # [i,j), sliding window indexes
        i, j, result = 0, 0, 0

        while j < l:
            if s[j] in hashtable and i < hashtable[s[j]] + 1:
                # slide i
                i = hashtable[s[j]] + 1

            # override
            hashtable[s[j]] = j
            j += 1

            # update result
            if result < j - i:
                result = j - i

        return result
